From 8c0604d1f086ef68a9d6f372e8ff041d1e063bec Mon Sep 17 00:00:00 2001
From: Natalia Kovalenko <natasha.k@variscite.com>
Date: Thu, 28 Mar 2024 14:31:51 +0100
Subject: [PATCH] bluetooth: BLE advertisement: Keep max_advt_instances values ​​aligned

Bluetooth stack should have the same way to check for BLE multi advertising
capability. Currently the Bluetooth native stack has this way to check
this capability:

from system/gd/hci/controller.cc:

bool is_supported(OpCode op_code) {
    switch (op_code) {
<many lines redacted>
    case OpCode::LE_MULTI_ADVT:
        return vendor_capabilities_.max_advt_instances_ != 0;
}

and the Bluethooth system service has this one instead:

from android/app/src/com/android/bluetooth/btservice/AdapterService.java

private static final int MIN_ADVT_INSTANCES_FOR_MA = 5;

private boolean isMultiAdvertisementSupported() {
    AdapterService service = getService();
    if (service == null) {
        return false;
    }

    int val = service.mAdapterProperties.getNumOfAdvertisementInstancesSupported();
    return val >= MIN_ADVT_INSTANCES_FOR_MA;
}

That's why android apk which checks this capability (like nRF connect) report:
"Multiple Advertisement supported NO"
but in native stack the code execution jumps into Multiple Advertisement support
implementation when the controller_->GetVendorCapabilities().max_advt_instances_
returns 1.

This misalignment causes a crash when we try to enable BLE Advertiser.

Keeping criteria unified when checking for BLE multi advertising capability
will solve the issue.

Change-Id: Id203746422fde0d794adb9d4afaa4f8cb3fad9e4
Signed-off-by: Natalia Kovalenko <natasha.k@variscite.com>
---
  system/gd/hci/controller.cc                                  | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletions(-)

diff --git a/system/gd/hci/controller.cc b/system/gd/hci/controller.cc
index c08ecab9d5..9db257c82c 100644
--- a/system/gd/hci/controller.cc
+++ b/system/gd/hci/controller.cc
@@ -36,6 +36,9 @@ constexpr uint8_t kMinEncryptionKeySize = 7;  // #define MIN_ENCRYPTION_KEY_SIZE
 constexpr bool kDefaultVendorCapabilitiesEnabled = true;
 static const std::string kPropertyVendorCapabilitiesEnabled =
     "bluetooth.core.le.vendor_capabilities.enabled";
+/* Keep this constant aligned with MIN_ADVT_INSTANCES_FOR_MA from
+ * android/app/src/com/android/bluetooth/btservice/AdapterService.java */
+static constexpr uint8_t min_advt_instances_for_ma = 5;
 
 using os::Handler;
 
@@ -1013,7 +1016,7 @@ struct Controller::impl {
       case OpCode::LE_GET_VENDOR_CAPABILITIES:
         return vendor_capabilities_.is_supported_ == 0x01;
       case OpCode::LE_MULTI_ADVT:
-        return vendor_capabilities_.max_advt_instances_ != 0x00;
+        return vendor_capabilities_.max_advt_instances_ >= min_advt_instances_for_ma;
       case OpCode::LE_BATCH_SCAN:
         return vendor_capabilities_.total_scan_results_storage_ != 0x00;
       case OpCode::LE_ADV_FILTER:
-- 
2.34.1
